### DISCUSSION TITLE
Elements flipped even though Newton solver converged

### DISCUSSION THREAD

**Discussion by tairoon1** [idaholab/moose#16682] [Q&A Modules: Solid mechanics] [answered]
Hello everyone,

I have been using MOOSE to compute some mechanics problems. I implemented a Neo-Hookean material model by defining a new strain function which sets the deformation gradient as the total strain, and a stress function which computes the 1st Piola Kirchhoff stress for a Neo-Hookean solid. The stiffness is derived using Automatic Differentiation. The files can be found [here](https://github.com/tairoon1/moosefiles).

Now, for small deformations, this yields good results and I can also observe a quadratic convergence of the Newton solver. But I noticed for a specific problem, that sometimes the Newton solver would converge, but the end result would contain flipped elements 
![Screenshot from 2021-01-13 17-57-30](https://user-images.githubusercontent.com/24324093/104483921-f3616a80-55c8-11eb-9813-8862cdf06d66.png). This example contains a uniaxial compression in the x direction of a square. The input and mesh file leading to this final deformation are also given in the link above. Using a smaller timestep size seems to prevent this problem from happening. However, I am confused, whether something with my code is wrong or what is going on?

I also used another FE solver and compared it with the MOOSE results. For "easy" simulations, both results are exactly the same, the residuals in each Newton iteration are exactly the same, but for more complicated case, MOOSE converges a lot slower... When I take a very small timestep in MOOSE dt=0.01 (element does not flip over), the same result is obtained as my other solver, which only needs a single time step dt=1. I also tried to not use Automatic Differentiation and directly code the consistent stiffness, however, it also lead to the same suboptimal convergence.

One difference in both solvers that I noticed is that MOOSE seems to use some line search algorithm that prevents the residual from increasing. In my other solver, that is not the case and always a full Jacobian step is used. Is there any way to deactivate the line search in MOOSE?

Can anyone help me with this? I am really confused and honestly frustrated that MOOSE sometimes flips over my elements without any indicator or warnings...

Thanks.

**Comment by hugary1995** [comment, marked as answer]
A few things you could try:
1. To turn off line search, you can add `line_search = none` in your executioner block.
2. For mechanics problems, if memory isn't a concern, few solvers can beat superlu in terms of speed. To use that, set
```
  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
  petsc_options_value = 'lu       superlu_dist                 '
```
3. Automatic scaling might also help: `automatic_scaling = true`
4. We've seen before that if you take a very large step, setting `preset = false` in your `FunctionDirichletBC`s can improve convergence and avoid spurious results.

**Reply by tairoon1** [reply]
Wow, number 4 did the magic! Thank you so much!

Do you mind telling me, what preset = false does?

**Reply by hugary1995** [reply]
You can take a look at the documentation for DirichletBC:
[https://mooseframework.inl.gov/source/bcs/DirichletBC.html](https://mooseframework.inl.gov/source/bcs/DirichletBC.html)
`preset = true` and `preset = false` correspond to two different ways of enforcing Dirichlet BC. Depending on your use case, one maybe better than the other.

**Comment by hugary1995** [comment]
Also, in response to your complaint

> I am really confused and honestly frustrated that MOOSE sometimes flips over my elements without any indicator or warnings

There is actually a very useful Userobject to do exactly that: [https://mooseframework.inl.gov/source/userobject/ElementQualityChecker.html](https://mooseframework.inl.gov/source/userobject/ElementQualityChecker.html)
which can be used together with an Auxkernel for visualization purposes:
[https://mooseframework.inl.gov/source/auxkernels/ElementQualityAux.html](https://mooseframework.inl.gov/source/auxkernels/ElementQualityAux.html)

**Reply by tairoon1** [reply]
I see. Thank you for the hint.
